### PR TITLE
Ml discussion

### PR DESCRIPTION
**Pull request by mlap**
Not completely content with the edits I've made but I think it's at a good enough stage to converse on where to go from here.

### CODE REVIEW

File: manuscript/manuscript.Rmd [hunk 1]
```diff
@@ -491,68 +491,51 @@ Many of our most realistic ecological models exist only as numerical simulations
...
-For a range of reasons, including high variance [@cite] and the exploration vs. exploitation trade-off [@cite], model-free and model-based algorithms tend to have unstable performance on environments with non-tabular state-action spaces [cite RL that matters].
-For instance, it is not uncommon to observe vastly different behavior when an agent is initialized with different random seeds [cite ibid].
-Such unstable behavior can often be avoided by alternate selection of hyperparameters, so researchers should consider stability as well as performance during hyperparameter selection.
-The choice of RL algorithm, design of the reward function, and duration of training can also influence stability.
-
+In seeking either to adopt or assess a deep RL-based approach to an ecological issue, we highlight three key issues that are important to bear in mind: generalizability, interpretability, and reward design.
+
+<!-- reason 1: generalizability-->
+<!-- ML: pivoting this to generalizability but still rewrote past paragraph which might fit in the appendix somewhere-->
+Currently, deep RL agents struggle to generalize their skills [@pmlr-v97-cobbe19a].
```
Comment (cboettig) [inline manuscript/manuscript.Rmd:503]: this is great but I think we need an example of what 'generalize their skills' means in an ecological context.  Or maybe just use language that is more concrete and less evocative of robots -- i.e. 'deep RL agents trained in one narrowly defined environment may struggle to perform well even on a closely related environment'?

File: manuscript/manuscript.Rmd [hunk 2]
```diff
@@ -491,68 +491,51 @@ Many of our most realistic ecological models exist only as numerical simulations
...
+This inability can affect performance during training and at evaluation. 
+For instance, training an agent on an environment that demands learning an assortment of skills is typically much more difficult than training an agent to learn one specific skill [@teh_distral_2017].
```
Comment (cboettig) [inline manuscript/manuscript.Rmd:505]: same as above, "skills" is not really a term we've defined.  Also worry about conflating two things here: this could sound like a transfer learning issue (does the policy learned in one environment hold true in other environments; an issue which is essentially independent from neural network/RL algo details) vs issues that are more specific to RL (i.e. from bias in bias-variance tradeoff) even within a single environment?

File: manuscript/manuscript.Rmd [hunk 3]
```diff
@@ -491,68 +491,51 @@ Many of our most realistic ecological models exist only as numerical simulations
...
+And, at evaluation, if you give an agent a novel task, including tasks that are closely related to the examples that the agent saw in training, the agent can easily fail to generalize its learned behavior to this new situation [@pmlr-v97-cobbe19a].
+These failures are all due to the brittleness of deep neural networks.
```
Comment (cboettig) [inline manuscript/manuscript.Rmd:507]: "all" might be a little strong? Isn't there an aspect of this that is due to the nature of a given environment than the network alone?

File: manuscript/manuscript.Rmd [hunk 4]
```diff
@@ -491,68 +491,51 @@ Many of our most realistic ecological models exist only as numerical simulations
...
+The flexibility of the approach makes it relatively straightforward to include realistic biology and environmental science in framing an RL problem, but training an RL agent to reach an acceptable level of performance can often be more challenging than expected.
 As the examples we have presented illustrate, it is important to begin with simple problems, including those for which an optimal strategy is already known.
 Such examples provide important benchmarks against which to calibrate the performance, tuning and training requirements of RL,
 and can easily be extended into more complex problems once RL agents can master the basics.
 Finding acceptable policies even in relatively simple problems may prove beyond the reach of existing algorithms and available computing power.
-<!-- Marcus is working on tightening up these next few paragraphs, precisely which issues actually get highlighted TBD-->
-In seeking either to adopt or assess a deep RL-based approach to an ecological issue, we highlight three key issues that are important to bear in mind: instability, interpretability, and reward design.
```
Comment (cboettig) [inline manuscript/manuscript.Rmd:500]: Seems like we still need to acknowledge instability as a fundamental issue?  I think the explicit observation that even the random seed impacts the outcomes in far more significant ways than it does in the many classical optimization algorithms that also use pseudo-random number, is an important concrete thing to mention that is kinda lost in the broader discussion of generalizability? 

I think the previous paragraph on instability wasn't perfect but identified concrete issues and had some concrete recommendations, the new paragraph on skills feels a lot more vague by comparison.

File: manuscript/manuscript.Rmd [hunk 5]
```diff
@@ -491,68 +491,51 @@ Many of our most realistic ecological models exist only as numerical simulations
...
-
 <!-- reason 2: interpretability -->
-<!-- ML: Re our OU talk, might be worth saying that this is a barrier to adoption? If we want to ultimately use these RL agents to make important decisions, it would be hard to swallow us blindly using them without having any sense to what is going on. -->
-Interpretability is a common challenge at the adoption of most ML methods [@cite]. 
-Although there have been some attempts to increase the transparency of RL agents that use neural network as function approximators [cite], it is still very difficult to explain the behavior of RL-trained agents.
-As our examples have illustrated, it may be possible to gain intuition on an agent's behavior by visualizing the action selected against each possible state, though this becomes more difficult in large problems.
-Another challenge is that Deep RL algorithms are very sample inefficient, all though progress is being made to improve this [@cite]. <!-- ML: I think we can cut this sentence, mentionned it in passing earlier-->
--->
-
+Deep RL agents are also difficult to interpret [@heuillet_explainability_2021].
```
Comment (cboettig) [inline manuscript/manuscript.Rmd:516]: This is definitely a good issue to discuss, but our examples don't illustrate it so well because we have 1-D action and 1-D state spaces where it's actually quite easy to visualize the policy and not all that hard to 'interpret' the policy.  It seems a bit incongruent to present some semi-interpretable examples.

File: manuscript/manuscript.Rmd [hunk 6]
```diff
@@ -491,68 +491,51 @@ Many of our most realistic ecological models exist only as numerical simulations
...
+Deep neural networks are commonly referred to as black box models, since it is challenging to fully explain why a deep neural network produces a specific output from a given input [@castelvecchi_can_2016].
+For deep RL, this means that we can rarely provide a meaningful explanation for why any agent selects an action.
+Compounded with the previously mentioned generalizability issues, the lack of explainability introduces a lot of safety concerns when attempting to apply these methods to the real world.
+When going from a simulator to the real world, the real world may introduce novel situations that a simulator-trained agent is not familiar with, and since we can not anticipate how an agent will respond to new situations, there is an inherent risk in letting RL agents make decisions in the real world.
```
Comment (cboettig) [inline manuscript/manuscript.Rmd:520]: I think this is basically a statement that any real-world use RL is really an exercise in transfer learning, since the real world isn't actually one of the environments it is trained on.  So the basic assumption is that interpretability might help intuition on when transfer learning will really transfer to the real world?

File: manuscript/manuscript.Rmd [hunk 7]
```diff
@@ -491,68 +491,51 @@ Many of our most realistic ecological models exist only as numerical simulations
...
-Moreover, specifying a reward function for conservation brings up various ethical concerns -- e.g. who gets to specify the reward function and will the values of the greater community be represented in the reward function.
-RL researchers often introduce modified reward functions during training that help an agent perform better on the original objective [@cite].
-This approach is particularly common when rewards are sparse, such as in a game of Chess where the only objective that counts is whether or not you win, but during training, an RL algorithm may reward the agent for intermediate steps such as capturing a piece.
+And lastly, a fundamental issue for all applied optimization problems is using the right objective.
+There have been many examples in which explicit reward design produces unanticipated agent behavior that is highly undesirable [@hadfield-menell_inverse_2020].
+Thus, for ecological systems, which are typically quite difficult to control, specifying a reward function could easily produce behavior that opposes the intents of the RL practitioner.
```
Comment (cboettig) [inline manuscript/manuscript.Rmd:525]: I think a concrete hypothetical would be useful here.  We haven't really discussed the idea of reward design as a strategy vs the reward really being the only thing you care about.  (e.g. if you're a fishing business and just trying to maximize profit, maybe you don't care about 'unexpected' solutions so long as it really is still maximizing the reward you care about.  The real problem comes along when the "RL" reward isn't the reward we care about, which isn't really a notion we have unpacked yet...

**Review by cboettig — CHANGES_REQUESTED** [review]
Nice work, but a few comments.  these three paragraphs are definitely some of the hardest.  I'm a bit worried they are just a bit vague.  I think they do a good job avoiding being too deep on the technical side, but are still very generic.  I think framing in terms of specific ecological/conservation scenarios (either from our examples or otherwise) could help clarify what we mean by "generalize", "interpret" and "reward design"

File: manuscript/manuscript.Rmd (hunk 2 above)
Comment (mlap) [reply manuscript/manuscript.Rmd:505]: I was going back and forth on using 'tasks' and 'skills' here -- I thought 'skills' might come off as more intuitive but will drop it. In RL literature, the issue that I talk about here is called multi-task learning. So I'll probably just keep to tasks and new environment language.

File: manuscript/manuscript.Rmd (hunk 3 above)
Comment (mlap) [reply manuscript/manuscript.Rmd:507]: Yeah, I intended 'all' as a qualifier on 'these', not as a modifier on 'due', will change this.

File: manuscript/manuscript.Rmd (hunk 4 above)
Comment (mlap) [reply manuscript/manuscript.Rmd:500]: Yeah, I took it out because I think of building more general agents as the bigger issue. Instability is a practical pain but if you find a good set of hyperparameters and an initialization seed then this problem is over. Building an agent that can generalize past experiences to new situations is a problem that has no solid answers to right now. I also think of generalization as a more likely failure mode than instability in terms of getting RL agents to inform decisions in the real world. Like if the agent is unstable in training, the consequence is that you will keep training until you get a good working agent. But if the agent fails to generalize to the real world, the consequence is that you could make horrible decisions because of this.

File: manuscript/manuscript.Rmd (hunk 5 above)
Comment (mlap) [reply manuscript/manuscript.Rmd:516]: For our models, yes interpretable but not explainable -- like it would still be hard to point inside the network and say what neuron is doing what. I'll probably just use 'transparency' to introduce.

File: manuscript/manuscript.Rmd (hunk 6 above)
Comment (mlap) [reply manuscript/manuscript.Rmd:520]: Yes, like if the agent does something drastic, it'll be hard to put trust in the agent without having a sense of why it's making that action.

Aside: I'm sensing you mean "generalization" in place of transfer learning here. Transfer learning would more so be if you are transferring a policy to an agent that you'd then train in another environment. If you're training and agent then evaluating it on a different environment, then it's a generalization problem.

**Comment by mlap** [comment]
@cboettig let me know what you think here. If we are desperate to cut words, I imagine we could fold the ethics paragraph into the RL open questions block and end on the paragraph we have after.

**Comment by cboettig** [comment]
@mlap I think this version looks great, and we should go ahead and merge. 

 I wanted to clean up the examples a little bit (where I think I can also trim word length), move the empirical example into the main text and then will have another look at the word count by section.
